File: cn_exim/public/js/purchase_invoice.js
frappe.ui.form.on("Purchase Invoice", {
    supplier: function (frm) {
        frm.set_query('custom_purchase_receipt', function () {
            return {
                filters: {
                    'supplier': frm.doc.supplier,
                    'docstatus': 1
                }
            };
        });
    },
    onload: function (frm) {
        if (frm.doc.docstatus == 0) {
            let has_purchase_order = (frm.doc.items || []).some(d => d.purchase_order && d.purchase_order.trim() !== '');

            if (has_purchase_order) {
                $.each(frm.doc.items || [], function (i, d) {
                    if (frm.doc.payment_terms_template == undefined && d.purchase_order != undefined) {
                        frappe.call({
                            method: "cn_exim.config.py.purchase_invoice.get_payment_trams",
                            args: {
                                name: d.purchase_order
                            },
                            callback: function (response) {

                                let data = response.message[0]
                                if (data.payment_terms_template) {
                                    frm.set_value("payment_terms_template", data['payment_terms_template'])
                                }

                            }
                        })
                    }
                })
            }
        }
        frm.set_query("custom_parent_cost_center", function() {
            return {
                "filters": {
                    "is_group": 1
                }
            };
        });

        frm.set_query("segment", function() {
            return {
                "filters": {
                    'parent_segment': frm.doc.custom_parent_cost_center,
                    "is_group": 0,
                    "disable": 0

                }
            };
        });

        console.log("Onload: Queries set for custom_cc and segment fields");
        // Set query for custom_cc field in the child table
        frm.fields_dict["items"].grid.get_field("custom_parent_cost_centre").get_query = function() {
            return {
                filters: {
                    is_group: 1
                }
            };
        };

        // Set query for custom_cost_centre field in the child table
        frm.fields_dict["items"].grid.get_field("segment").get_query = function(doc, cdt, cdn) {
            var child = locals[cdt][cdn];
            return {
                filters: {
                    parent_segment: child.custom_parent_cost_centre,
                    is_group: 0,
                    disable: 0
                }
            };
        };
    },

    refresh: frm => {
        if (frm.doc.items && frm.doc.docstatus == 0) {
            if (frm.doc.items[0].purchase_receipt) {
                frappe.call({
                    method: 'cn_exim.config.py.purchase_invoice.get_due_date_based_on_condition',
                    args: { pr: frm.doc.items[0].purchase_receipt },
                    freeze: true,
                    callback: r => {
                        if (r.message.due_date) {
                            frm.set_df_property('cnp_section_break_wob2z', 'hidden', true);
                        }
                        else if (r.message.status) {
                            frm.set_df_property('cnp_section_break_wob2z', 'hidden', false);
                        }
                    }
                })
            }
        }
        if (frm.doc.docstatus === 1) {
            frm.add_custom_button("Landed Cost Voucher", function () {
                if (!frm.doc.custom_purchase_order) {
                    let po_list = [];
                    let all_purchase_receipts = [];
                    let all_expenses = [];
                    let all_purchase_invoices = [];

                    frm.doc.items.forEach(element => {
                        if (element.custom_po_number) {
                            if (!po_list.includes(element.custom_po_number)) {
                                po_list.push(element.custom_po_number);
                            }

                            all_expenses.push({
                                expense_account: element.expense_account,
                                description: element.description,
                                amount: element.amount,
                            });

                        }
                    });
                    frm.doc.taxes.forEach(element => {
                        all_expenses.push({
                            expense_account: element.account_head,
                            description: element.description,
                            amount: element.tax_amount,
                        });
                    })

                    // Add Purchase Invoice (only one)
                        all_purchase_invoices.push({
                            purchase_invoice: frm.doc.name,
                            supplier: frm.doc.supplier,
                            amount: frm.doc.grand_total,
                            posting_date: frm.doc.posting_date,
                        });
                    // Correct way: loop through Object.entries()


                    po_list.forEach(po_number => {
                        frappe.call({
                            method: "cn_exim.config.py.purchase_invoice.get_landed_cost_voucher_details",
                            args: {
                                purchase_invoice_name: frm.doc.name,
                                custom_purchase_order: po_number
                            },
                            callback: function (r) {
                                const data = r.message; // Assuming this is your response

                                // Add Purchase Receipt
                                if (data[0] && data[0].length) {
                                    data[0].forEach(pr => {
                                        // Check if the receipt already exists in all_purchase_receipts
                                        const exists = all_purchase_receipts.some(item => item.receipt_document === pr.name);
                                        if (!exists) {
                                            all_purchase_receipts.push({
                                                receipt_document_type: 'Purchase Receipt',
                                                receipt_document: pr.name,
                                            });
                                        }
                                    });
                                }

                                // After processing all PO numbers, create Landed Cost Voucher
                                if (Object.entries(po_list).length === Object.keys(all_purchase_receipts).length) {
                                    // Create the Landed Cost Voucher once all POs are processed
                                    frappe.call({
                                        method: "frappe.client.insert",
                                        args: {
                                            doc: {
                                                doctype: "Landed Cost Voucher",
                                                naming_series: "LCV-.YYYY.-",
                                                company: frm.doc.company, // put your company
                                                posting_date: frappe.datetime.get_today(),
                                                distribute_charges_based_on: "Amount",
                                                purchase_receipts: all_purchase_receipts,
                                                taxes: all_expenses,
                                                custom_purchase_invoice_details: all_purchase_invoices,
                                            }
                                        },
                                        callback: function(r) {
                                            if (!r.exc) {
                                                frappe.msgprint(`Landed Cost Voucher ${r.message.name} created`);
                                                frappe.set_route("Form", "Landed Cost Voucher", r.message.name);
                                            }
                                        }
                                    });
                                }
                            }
                        });
                    });

                }
                else {
                    frappe.call({
                        method: "cn_exim.config.py.purchase_invoice.get_landed_cost_voucher_details",
                        args: {
                            purchase_invoice_name: frm.doc.name,
                            // purchase_invoice_items:frm.doc.items,
                            custom_purchase_order: frm.doc.custom_purchase_order
                        },
                        callback: function (r) {
                            if (r.exc) {
                                frappe.throw("Error: " + r.exc);
                                return;
                            }
                            let purchase_receipts = r.message[0]
                            let taxes_data = r.message[1] || [];
                            let purchase_invoice_details = r.message[2] || []
                            // purchase_receipts.forEach(element => {

                            // });

                            let seen_receipts = new Set();
                            let purchase_receipt_rows = [];

                            if (Array.isArray(purchase_receipts)) {
                                purchase_receipts.forEach(receipt => {
                                    if (!seen_receipts.has(receipt.name)) {
                                        seen_receipts.add(receipt.name);
                                        purchase_receipt_rows.push({
                                            receipt_document_type: 'Purchase Receipt',
                                            receipt_document: receipt.name,
                                            supplier: receipt.supplier,
                                            grand_total: receipt.grand_total
                                        });
                                    }
                                });
                            }

                            // If it's a single object
                            // else if (typeof purchase_receipts === 'object') {
                            //     purchase_receipt_rows = [{
                            //         receipt_document_type: 'Purchase Receipt',
                            //         receipt_document: purchase_receipts.name,
                            //         supplier: purchase_receipts.supplier,
                            //         grand_total: purchase_receipts.grand_total
                            //     }];
                            // }

                            let taxes_rows = [];
                            if (Array.isArray(taxes_data)) {
                                taxes_rows = taxes_data.map(tax => {
                                    return {
                                        expense_account: tax.account_head,
                                        description: tax.description,
                                        amount: tax.amount
                                    };
                                });
                            }
                            // Store purchase invoice details for context (will be used for custom fields if needed)
                            let pi_details_rows = [];
                            if (Array.isArray(purchase_invoice_details) && purchase_invoice_details.length > 0) {
                                pi_details_rows = purchase_invoice_details.map(pi => {
                                    return {
                                        purchase_invoice: pi.purchase_invoice,
                                        expense_account: pi.expense_account,
                                        amount: pi.amount,
                                        supplier: pi.supplier,
                                        date: pi.date,
                                        item: pi.item,
                                    }
                                })
                            }

                            if (purchase_receipt_rows.length > 0) {
                                frappe.call({
                                    method: "frappe.client.insert",
                                    args: {
                                        doc: {
                                            'doctype': "Landed Cost Voucher",
                                            'distribute_charges_based_on': "Amount",
                                            'purchase_receipts': purchase_receipt_rows,
                                            'taxes': taxes_rows,
                                            'custom_purchase_invoice_details': pi_details_rows
                                        }
                                    },
                                    callback: function (r) {

                                        if (!r.exc) {
                                            frappe.set_route("Form", "Landed Cost Voucher", r.message.name);
                                        }
                                    }
                                })
                            } else {
                                frappe.msgprint("No Purchase Receipts found for this Purchase Order.");
                            }
                        }
                    });
                }

            }, __("Create"))
        }



        frm.add_custom_button("BOE Entry", function () {
            var d = new frappe.ui.form.MultiSelectDialog({
                doctype: "BOE Entry",
                target: this.cur_frm,
                setters: {
                    vendor: null,
                    boe_date: null,
                },
                add_filters_group: 1,
                columns: ["name", "boe_date", "vendor"],
                get_query() {
                    return {
                        filters: [
                            ["docstatus", "=", 1],
                            ["status", "!=", "To Bill"]
                        ]
                    };
                },
                action(selections) {
                    var length = selections.length
                    if (length > 0) {
                        frappe.call({
                            method: "cn_exim.config.py.purchase_invoice.get_invoice_data_for_import",
                            args: {
                                boe_list: selections
                            },
                            callback: function (r) {
                                frm.set_value("supplier", r.message[1])
                                frm.set_value("currency", r.message[2])

                                frm.clear_table("items");

                                row_list = r.message[0]
                                row_list.forEach(obj => {
                                    let row = frm.add_child("items");

                                    frappe.call({
                                        method: "frappe.client.get_value",
                                        args: {
                                            doctype: "Item",
                                            filters: {
                                                name: obj.item_code
                                            },
                                            fieldname: ["item_name", "stock_uom"]   // ✅ (not 'filed', should be 'fieldname')
                                        },
                                        callback: function (r) {
                                            if (r.message) {
                                                row.item_name = r.message.item_name;
                                                row.uom = r.message.stock_uom;
                                                frm.refresh_field("items");
                                            }
                                        }
                                    });

                                    // Set other fields
                                    row.custom_po_number = obj.custom_po_number;
                                    row.custom_boe_entry = obj.custom_boe_entry;
                                    row.item_code = obj.item_code;
                                    row.rate = obj.rate;
                                    row.amount = obj.amount;
                                    row.description = obj.description;
                                    row.qty = obj.qty;
                                })
                                frm.refresh_field("items")
                            }
                        })
                    }
                    d.dialog.hide();
                }
            });
        }, __("Get Items From"))

        // Function to generate table rows for each item in the Purchase Invoice
        function generateItemRows(items) {
            let rows = '';
            if (items && items.length > 0) {
                items.forEach(function(item) {
                    rows += `
                    <tr>
                        <td>${item.item_code}</td>
                        <td>${item.expense_head || item.expense_account || 'N/A'}</td>
                        <td>${item.qty}</td>
                        <td>${item.rate}</td>
                        <td>${item.amount}</td>

                    </tr>
                    `;
                });
            } else {
                rows += `
                <tr>
                    <td colspan="5">No items available.</td>
                </tr>
                `;
            }
            return rows;
        }

        // Function to generate GST rows
        function generateGSTRows(taxes) {
            let rows = '';
            if (taxes && taxes.length > 0) {
                taxes.forEach(function(tax) {
                    rows += `
                    <tr>
                        <td>${tax.account_head}</td>
                        <td>${tax.rate || '0.00'}%</td>
                        <td>${tax.tax_amount || '0.00'}</td>
                        <td>${tax.description || 'N/A'}</td>
                    </tr>
                    `;
                });
            } else {
                rows += `
                <tr>
                    <td colspan="4">No GST details available.</td>
                </tr>
                `;
            }
            return rows;
        }

        // Function to generate the summary rows
        function generateSummaryRows(doc) {
            return `
            <tr>
                <td colspan="4" style="text-align:right;"><strong>Total Amount:</strong></td>
                <td>${doc.total || '0.00'}</td>
            </tr>
            <tr>
                <td colspan="4" style="text-align:right;"><strong>Total Taxes and Charges:</strong></td>
                <td>${doc.total_taxes_and_charges || '0.00'}</td>
            </tr>
            <tr>
                <td colspan="4" style="text-align:right;"><strong>Grand Total:</strong></td>
                <td>${doc.grand_total || '0.00'}</td>
            </tr>
            <tr>
                <td colspan="6" style="text-align:left;"><strong>Amount in Words:</strong> ${doc.in_words || 'Zero'}</td>
            </tr>
            `;
        }

        // Main HTML content for the custom Purchase Invoice section
        let table_html = `
        <style>
            .custom-table {
                width: 100%;
                border-collapse: collapse;
                margin-top: 10px;
            }
            .custom-table th, .custom-table td {
                padding: 10px;
                text-align: left;
                border: 1px solid #ddd;
            }
            .custom-table th {
                background-color: #f4f4f4;
                color: #333;
                font-weight: bold;
            }
            .custom-table tr:nth-child(even) {
                background-color: #f9f9f9;
            }
            .custom-table tr:hover {
                background-color: #e0e0e0;
            }
            .table-title {
                margin-top: 20px;
                font-size: 16px;
                font-weight: bold;
                text-align: left;
            }
            .custom-invoice-details {
    display: flex;
    flex-wrap: wrap;
    gap: 10px;
}

.custom-invoice-detail {
    flex: 1;
    min-width: 200px;
    padding: 8px;
    background-color: white;
    border: 1px solid #ddd;
    border-radius: 5px;
    box-shadow: 0 2px 4px rgba(0,0,0,0.1);
}

        </style>

<div class="custom-invoice-info">
    <h3>Invoice Summary</h3>
    <div class="custom-invoice-details">
        <div class="custom-invoice-detail">
            <p><strong>Invoice Number:</strong> ${frm.doc.name || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Date:</strong> ${frm.doc.posting_date || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Created By:</strong> ${frm.doc.owner || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Workflow State:</strong> ${frm.doc.workflow_state || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Supplier:</strong> ${frm.doc.supplier || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Tally Invoice No:</strong> ${frm.doc.tally_invoice_number_refrence || 'N/A'}</p>
        </div>

        <div class="custom-invoice-detail">
            <p><strong>Supplier Invoice No:</strong> ${frm.doc.bill_no || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Supplier Invoice Date:</strong> ${frm.doc.bill_date || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Segment:</strong> ${frm.doc.segment || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Cost Center:</strong> ${frm.doc.cost_center || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Billing Address:</strong> ${frm.doc.billing_address_display || 'N/A'}</p>
        </div>
        <div class="custom-invoice-detail">
            <p><strong>Shipping Address:</strong> ${frm.doc.shipping_address_display || 'N/A'}</p>
        </div>
    </div>
</div>


        <h3 class="table-title">Item Details</h3>
        <table class="custom-table">
            <thead>
                <tr>
                    <th>Item Code</th>
                    <th>Expense Head</th>
                    <th>Quantity</th>
                    <th>Rate</th>
                    <th>Amount</th>

                </tr>
            </thead>
            <tbody>
                ${generateItemRows(frm.doc.items)}
                ${generateSummaryRows(frm.doc)}
            </tbody>
        </table>

        <h3 class="table-title">GST Details</h3>
        <p><strong>Taxes and Charges Template:</strong> ${frm.doc.taxes_and_charges || 'N/A'}</p>
        <table class="custom-table">
            <thead>
                <tr>
                    <th>Account Head</th>
                    <th>Rate (%)</th>
                    <th>Tax Amount</th>
                    <th>Description</th>
                </tr>
            </thead>
            <tbody>
                ${generateGSTRows(frm.doc.taxes)}
            </tbody>
        </table>
        `;

        // Insert the generated HTML into the custom field
        frm.fields_dict.custom_custom_table.$wrapper.html(table_html);
    },
    before_save: frm => {
        let is_block = false
        $.each(frm.doc.payment_schedule || [], function (i, d) {
            if (d.custom_purchase_invoice_blocked) {
                is_block = true
            }
        })

        if (is_block) {
            frm.set_value("on_hold", 1)
        }
        if (frm.doc.items) {
            if (frm.doc.items[0].purchase_receipt) {
                frappe.call({
                    method: 'cn_exim.config.py.purchase_invoice.get_due_date_based_on_condition',
                    args: { pr: frm.doc.items[0].purchase_receipt },
                    freeze: true,
                    callback: r => {
                        if (r.message.due_date) {
                            frm.doc.payment_schedule[0].due_date = r.message.due_date;
                            frm.doc.due_date = r.message.due_date;
                        }
                        else if (r.message.status) {
                            let installation_date = frm.doc.cnp_installation_date
                            let expected_installation_date = frm.doc.cnp_expected_installation_date
                            if (frm.doc.cnp_is_installed && installation_date) {
                                frm.doc.payment_schedule[0].due_date = installation_date;
                                frm.doc.due_date = installation_date;
                            }
                            else if (!frm.doc.cnp_is_installed && expected_installation_date) {
                                frm.doc.payment_schedule[0].due_date = expected_installation_date;
                                frm.doc.due_date = expected_installation_date;
                            }
                        }
                    }
                })
            }
        }
    },
    cnp_expected_installation_date: frm => {
        update_due_date(frm)
    },
    cnp_installation_date: frm => {
        update_due_date(frm)
    },
    on_update: frm => {
    },
    on_submit:function(frm){
        boe_list = []
        $.each(frm.doc.items || [], function (i, d) {
            if (d.custom_boe_entry) {
                boe_list.push(d.custom_boe_entry)
            }
        })
        if (boe_list.length > 0) {
            frappe.call({
                method: "cn_exim.config.py.purchase_invoice.update_boe_entry",
                args: {
                    boe_list: boe_list,
                },
                callback: function (r) {
                    if (!r.exc) {
                        frappe.msgprint("BOE Entry updated successfully");
                    }
                }
            })
        }
    },
    item_group: function(frm) {
        frm.set_query("segment", function() {
            return {
                "filters": {
                    'parent_segment': frm.doc.custom_parent_cost_center,
                    "is_group": 0,
                    "disable": 0
                }
            };
        });
    },
    custom_parent_cost_center: function(frm, cdt, cdn) {
        frm.set_query("segment", function() {
            return {
                "filters": {
                    'parent_segment': frm.doc.custom_parent_cost_center,
                    "is_group": 0,
                    "disable": 0
                }
            };
        });

        var child = locals[cdt][cdn];
        frappe.ui.form.on('Purchase Invoice Item', {
            custom_parent_cost_centre: function(frm, cdt, cdn) {
                var child = locals[cdt][cdn];
                frm.fields_dict["items"].grid.get_field("segment").get_query = function() {
                    return {
                        filters: {
                            parent_segment: child.custom_parent_cost_centre,
                            is_group: 0,
                            disable: 0
                        }
                    };
                };
            }
        });
    }
});

function update_due_date(frm) {
    let installation_date = frm.doc.cnp_installation_date
    let expected_installation_date = frm.doc.cnp_expected_installation_date
    if (frm.doc.cnp_is_installed && installation_date) {
        frm.doc.payment_schedule[0].due_date = installation_date;
        frm.doc.due_date = installation_date;
    }
    else if (!frm.doc.cnp_is_installed && expected_installation_date) {
        frm.doc.payment_schedule[0].due_date = expected_installation_date;
        frm.doc.due_date = expected_installation_date;
    }
}



frappe.ui.form.on('Purchase Invoice Item', {
    item_tax_template: function(frm, cdt, cdn) {
        let item = locals[cdt][cdn];
        let rate = item.rate;
        let item_tax_template = item.item_tax_template;

        if (item_tax_template) {
            frappe.db.get_doc('Item Tax Template', item_tax_template)
                .then(doc => {
                    if (doc && doc.gst_rate) {
                        let gst_rate = doc.gst_rate;
                        let gst_rate_with_percentage = `${gst_rate}%`;
                        let gst_rate_number = parseFloat(gst_rate);
                        let calculated_value = (rate * gst_rate_number) / 100;
                        frappe.model.set_value(cdt, cdn, 'custom_calculated_value', calculated_value);

                        // Check if the 'custom_gst_not_receivable' checkbox is checked
                        if (item.custom_gst_not_receivable) {
                            // Add the calculated value to the rate
                            let new_rate = rate + calculated_value;
                            frappe.model.set_value(cdt, cdn, 'rate', new_rate);
                            frappe.model.set_value(cdt, cdn, 'item_tax_template', '');
                        }

                        console.log('Rate:', rate);
                        console.log('GST Rate with Percentage:', gst_rate_with_percentage);
                        console.log('Calculated Value:', calculated_value);
                    }
                });
        } else {
            console.log('Rate:', rate);
            console.log('GST Rate with Percentage: No Tax Template');
        }
    },

    custom_gst_not_receivable: function(frm, cdt, cdn) {
        let item = locals[cdt][cdn];
        let rate = item.rate;
        let calculated_value = item.custom_calculated_value;

        // Check if the checkbox is checked
        if (item.custom_gst_not_receivable) {
            // Add the calculated value to the rate
            let new_rate = rate + calculated_value;
            frappe.model.set_value(cdt, cdn, 'rate', new_rate);
            frappe.model.set_value(cdt, cdn, 'item_tax_template', '');
        }
    }
});

File: cn_exim/config/py/purchase_invoice.py
import frappe
from frappe.utils import now_datetime, add_days, formatdate

@frappe.whitelist()
def get_payment_trams(name):
    data = frappe.db.sql(" select payment_terms_template from `tabPurchase Order` where name=%s ", (name), as_dict=True)
    
    return data


@frappe.whitelist()
def get_due_date_based_on_condition(pr):
    purchase_receipt = frappe.get_doc("Purchase Receipt", pr)
    if purchase_receipt.items[0].purchase_order:
        purchase_order = frappe.get_doc("Purchase Order", purchase_receipt.items[0].purchase_order)
    
        payment_term = None
        if purchase_order.payment_schedule:
            payment_term = purchase_order.payment_schedule[0].payment_term
        
        if payment_term:
            due_date_based_on = frappe.get_value("Payment Term", payment_term, "due_date_based_on")
            days_to_add = frappe.get_value("Payment Term", payment_term, "credit_days")
            
            if due_date_based_on == "After invoice creation":
                today_datetime = now_datetime()
                new_date = add_days(today_datetime, days_to_add)
                return {"due_date":frappe.utils.formatdate(new_date.date(), "yyyy-mm-dd")}
            elif due_date_based_on == "After PR is created":
                today_datetime = purchase_receipt.posting_date
            else:
                # Handle the case when due_date_based_on is not in either dictionary
                if due_date_based_on in ["Completion of Work","On Installation"]:
                    return {"status":True}
                return None
            
            new_date = add_days(today_datetime, days_to_add)
            return {"due_date":new_date}

        # Handle the case when payment_term is not found
        return None

@frappe.whitelist()
def get_landed_cost_voucher_details(purchase_invoice_name, custom_purchase_order):
    # print("@@@@@@@@@@-->",custom_purchase_order)
    # if custom_purchase_order:
    #     pass

    # purchase_order_item=frappe.db.sql("select name from `tabPurchase Order Item` where parent=%s",[custom_purchase_order],as_dict=True)

    # purchase_order_item_name=purchase_order_item[0]["name"]

    # items = frappe.get_all("Purchase Receipt Item",filters={"purchase_order_item":purchase_order_item_name},fields=["parent"])
    # print("@@@@@@@@@@@@@@@@@@@@@@@@->",items)
    # pi_doc = frappe.get_doc("Purchase Invoice", purchase_invoice_name)
    # pr_docs = []
    # for items in pi_doc.items:
    #     pr = items.parent
    #     pr_docs.append(pr)
        

    # pr_uni = set(pr_docs)
    # print("@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@",pr_uni)
    # return pr_uni

    try:
        purchase_receipts = []
        taxes_data = []
        purchase_invoice_details=[]
        
        # Check if we have a valid purchase order
        if not custom_purchase_order:
            frappe.throw(("Purchase Order reference is missing"))
        
        if purchase_invoice_name:
            pi_doc=frappe.get_doc("Purchase Invoice",purchase_invoice_name)
            purchase_invoice_details.append({
                "purchase_invoice":pi_doc.name,
                "supplier":pi_doc.supplier,
                "amount":pi_doc.total,
                "date":pi_doc.posting_date,
            })
            for item in pi_doc.items:
                taxes_data.append({
                    "account_head":item.expense_account,
                    "amount":item.amount,
                    "description":item.description,
                })
            for tax in pi_doc.taxes:
                taxes_data.append({
                    "charge_type": tax.charge_type,
                    "account_head": tax.account_head,
                    "description": tax.description,
                    "rate": tax.rate,
                    "amount": tax.tax_amount,
                    "total": tax.total,
                    "receipt_document": pi_doc.name
                })
        # Step 1: Get Gate Entry Details for this Purchase Order
        gate_entries = frappe.db.sql("""
            SELECT parent as gate_entry
            FROM `tabGate Entry Details`
            WHERE purchase_order = %s
        """, custom_purchase_order, as_dict=True)
        
        if not gate_entries:
            frappe.msgprint(("No Gate Entries found for Purchase Order {0}").format(custom_purchase_order))
            return [[]]
    
    
        # Step 2: For each Gate Entry, get the linked Purchase Receipts
        for ge in gate_entries:
            gate_entry_no=ge.gate_entry

            pr_list=frappe.db.sql("""
                SELECT name, supplier, grand_total
                FROM `tabPurchase Receipt`
                WHERE custom_gate_entry_no = %s
            """, gate_entry_no, as_dict=True)


            for pr in pr_list:
                purchase_receipts.append({
                    "name": pr.name,
                    "supplier": pr.supplier,
                    "grand_total": pr.grand_total
                })

                # pr_taxes = frappe.db.sql("""
                #         SELECT charge_type, account_head, description, rate, tax_amount, total
                #         FROM `tabPurchase Taxes and Charges`
                #         WHERE parent = %s
                #     """, pr.name, as_dict=True)
                
                # for tax in pr_taxes:
                #     taxes_data.append({
                #         "charge_type": tax.charge_type,
                #         "account_head": tax.account_head,
                #         "description": tax.description,
                #         "rate": tax.rate,
                #         "amount": tax.tax_amount,
                #         "total": tax.total,
                #         "receipt_document": pr.name
                #     })
        return [purchase_receipts, taxes_data, purchase_invoice_details]
    except Exception as e:
        frappe.log_error(f"Failed to get Landed Cost Voucher details: {str(e)}", 
                        "get_landed_cost_voucher_details error")
        frappe.throw(str(e))



@frappe.whitelist()
def get_invoice_data_for_import(boe_list):
    boe_list = frappe.json.loads(boe_list)

    # Store mapping between per_alert_check and boe_entry_no
    alert_to_boe_map = {}
    for boe_entry_no in boe_list:
        boe_doc = frappe.get_doc("BOE Entry", boe_entry_no)
        per_alert_check = boe_doc.get("per_alert_check")
        vendor = boe_doc.get("vendor")
        currency = boe_doc.get("currency")
        if per_alert_check:
            alert_to_boe_map[per_alert_check] = boe_entry_no

    child_details_list = []
    po_data_map = {}

    fields_to_sum = [
        "bcd_amount", "hcs_amount", "swl_amount", "total_duty",
        "freight_amount", "misc_charge_amt", "insurance_amount"
    ]

    # Add boe_entry_no to each item row
    for per_alert, boe_entry_no in alert_to_boe_map.items():
        data = frappe.db.sql(
            "SELECT * FROM `tabPre-Alert Item Details` WHERE parent = %s",
            (per_alert,),
            as_dict=True
        )
        for item in data:
            item["boe_entry_no"] = boe_entry_no
            child_details_list.append(item)

    # Aggregate values by po_no
    for item in child_details_list:
        po_no = item.get("po_no")
        boe_entry_no = item.get("boe_entry_no")
        if not po_no:
            continue

        key = (po_no, boe_entry_no)  # Combine po_no and boe_entry_no to uniquely group

        if key not in po_data_map:
            po_data_map[key] = {field: item.get(field, 0) or 0 for field in fields_to_sum}
            po_data_map[key]["po_no"] = po_no
            po_data_map[key]["boe_entry_no"] = boe_entry_no
        else:
            for field in fields_to_sum:
                po_data_map[key][field] += item.get(field, 0) or 0

    # Final conversion
    field_to_description = {
        "bcd_amount": "Basic Customs Duty",
        "hcs_amount": "Hazard Communication Standard",
        "swl_amount": "social welfare surcharge",
        "total_duty": "Duty",
        "freight_amount": "Freight",
        "misc_charge_amt": "Misc. Charge Amount",
        "insurance_amount": "Insurance Amount"
    }

    final_list = []

    for (po_no, boe_entry_no), values in po_data_map.items():
        for field, description in field_to_description.items():
            amount = values.get(field, 0) or 0
            if amount > 0:
                final_list.append({
                    "custom_po_number": po_no,
                    "custom_boe_entry": boe_entry_no,
                    "item_code": description,
                    "rate": amount,
                    "amount":amount,
                    "description": field,
                    "qty":1
                })

    # Example printout
    return final_list, vendor, currency


@frappe.whitelist()
def update_boe_entry(boe_list):
    boe_list = frappe.json.loads(boe_list)
    for boe_entry_no in boe_list:
        boe_doc = frappe.get_doc("BOE Entry", boe_entry_no)
        if boe_doc:
            boe_doc.update({
                "status": "To Bill",
            })
            boe_doc.save()
            frappe.db.commit()
